docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+
+    /**
+     * Event emitted when a trade offer is cancelled.
+     */
+    event OfferCancelled(bytes32 hash);
+
+    /**
+     * Event emitted when the public sale begins.
+     */
+    event SaleBegins();
+
+    /**
+     * Event emitted when the community grant period ends.
+     */
+    event CommunityGrantEnds();
+
+    bytes4 internal constant MAGIC_ON_ERC721_RECEIVED = 0x150b7a02;
+
+    // IPFS Hash to the NFT content
+    string public contentHash = "QmfXYgfX1qNfzQ6NRyFnupniZusasFPMeiWn5aaDnx7YXo";
+
+    uint public constant TOKEN_LIMIT = 20000;
+    uint public constant SALE_LIMIT = 9000;
+
+    mapping(bytes4 => bool) internal supportedInterfaces;
+
+    mapping (uint256 => address) internal idToOwner;
+
+    mapping (uint256 => uint256) public creatorNftMints;
+
+    mapping (uint256 => address) internal idToApproval;
+
+    mapping (address => mapping (address => bool)) internal ownerToOperators;
+
+    mapping(address => uint256[]) internal ownerToIds;
+
+    mapping(uint256 => uint256) internal idToOwnerIndex;
+
+    string internal nftName = "Meebits";
+    string internal nftSymbol = unicode"⚇";
+
+    uint internal numTokens = 0;
+    uint internal numSales = 0;
+
+    // Cryptopunks contract
+    address internal punks;
+
+    // Autoglyphs contract
+    address internal glyphs;
+
+    address payable internal deployer;
+    address payable internal beneficiary;
+    bool public communityGrant = true;
+    bool public publicSale = false;
+    uint private price;
+    uint public saleStartTime;
+    uint public saleDuration;
+
+    //// Random index assignment
+    uint internal nonce = 0;
+    uint[TOKEN_LIMIT] internal indices;
+
+    //// Market
+    bool public marketPaused;
+    bool public contractSealed;
+    mapping (address => uint256) public ethBalance;
+    mapping (bytes32 => bool) public cancelledOffers;
+
+    modifier onlyDeployer() {
+        require(msg.sender == deployer, "Only deployer.");
+        _;
+    }
+
+    bool private reentrancyLock = false;
+
+    /* Prevent a contract function from being reentrant-called. */
+    modifier reentrancyGuard {
+        if (reentrancyLock) {
+            revert();
+        }
+        reentrancyLock = true;
+        _;
+        reentrancyLock = false;
+    }
+
+    modifier canOperate(uint256 _tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(tokenOwner == msg.sender || ownerToOperators[tokenOwner][msg.sender], "Cannot operate.");
+        _;
+    }
+
+    modifier canTransfer(uint256 _tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(
+            tokenOwner == msg.sender
+            || idToApproval[_tokenId] == msg.sender
+            || ownerToOperators[tokenOwner][msg.sender], "Cannot transfer."
+        );
+        _;
+    }
+
+    modifier validNFToken(uint256 _tokenId) {
+        require(idToOwner[_tokenId] != address(0), "Invalid token.");
+        _;
+    }
```

Comment:
Function modifier to ensure the tokenid is a valid for meebits. Checks if there is an owner